File: myFinance_improved_V2/lib/features/auth/data/repositories/store_repository_impl.dart
// lib/features/auth/data/repositories/store_repository_impl.dart

import '../../domain/entities/store_entity.dart';
import '../../domain/repositories/store_repository.dart';
import '../datasources/supabase_store_datasource.dart';
import '../models/freezed/store_dto_mapper.dart';
import 'base_repository.dart';

/// Store Repository Implementation
///
/// 📜 계약 이행자 - Domain Repository Interface를 구현
///
/// 책임:
/// - Domain 계약 준수
/// - DataSource 호출
/// - Model ↔ Entity 변환 (operational settings 포함!)
/// - Exception 처리 및 변환
class StoreRepositoryImpl extends BaseRepository implements StoreRepository {
  final StoreDataSource _dataSource;

  StoreRepositoryImpl(this._dataSource);

  @override
  Future<Store> create(Store store) {
    return execute(() async {
      // Convert Entity to insert map (without ID)
      final insertMap = store.toInsertMap();

      // Call DataSource with operational settings
      final createdDto = await _dataSource.createStore(insertMap);

      // Convert DTO back to Entity
      return createdDto.toEntity();
    });
  }

  @override
  Future<Store?> findById(String storeId) {
    return executeNullable(() async {
      final model = await _dataSource.getStoreById(storeId);
      return model?.toEntity();
    });
  }

  @override
  Future<List<Store>> findByCompany(String companyId) {
    return execute(() async {
      final models = await _dataSource.getStoresByCompanyId(companyId);
      return models.map((m) => m.toEntity()).toList();
    });
  }

  @override
  Future<bool> codeExists({
    required String companyId,
    required String storeCode,
  }) {
    return execute(() async {
      return await _dataSource.isStoreCodeExists(
        companyId: companyId,
        storeCode: storeCode,
      );
    });
  }

  @override
  Future<Store> update(Store store) {
    return execute(() async {
      final updateData = {
        'store_name': store.name,
        'store_code': store.storeCode,
        'store_address': store.address,
        'store_phone': store.phone,
        'huddle_time': store.huddleTimeMinutes,
        'payment_time': store.paymentTimeDays,
        'allowed_distance': store.allowedDistanceMeters?.toInt(),
      };

      final updatedDto = await _dataSource.updateStore(
        store.id,
        updateData,
      );

      return updatedDto.toEntity();
    });
  }

  @override
  Future<void> delete(String storeId) {
    return execute(() async {
      await _dataSource.deleteStore(storeId);
    });
  }
}

File: website/src/features/store-setting/data/repositories/StoreRepositoryImpl.ts
import { IStoreRepository, StoreResult } from '../../domain/repositories/IStoreRepository';
import { Store } from '../../domain/entities/Store';
import { StoreDataSource } from '../datasources/StoreDataSource';

/**
 * StoreRepositoryImpl - IStoreRepository 구현체
 * Clean Architecture의 Data Layer에서 Domain Layer의 인터페이스를 구현
 * DataSource와 Domain Entity 사이의 매핑을 담당
 */
export class StoreRepositoryImpl implements IStoreRepository {
  private dataSource = new StoreDataSource();

  /**
   * 회사의 모든 활성 상점 조회
   */
  async getStores(companyId: string): Promise<StoreResult> {
    try {
      const data = await this.dataSource.getStores(companyId);

      // DTO → Domain Entity 변환
      // Backup과 동일한 필드 매핑: store_id, store_name, store_address, store_phone, huddle_time, payment_time, allowed_distance
      const stores = data.map((dto: any) =>
        new Store(
          dto.store_id,
          dto.company_id,
          dto.store_name,
          dto.store_address, // backup uses store_address
          dto.store_phone,   // backup uses store_phone
          !dto.is_deleted,   // is_active = NOT is_deleted
          dto.huddle_time,   // backup uses huddle_time
          dto.payment_time,  // backup uses payment_time
          dto.allowed_distance // backup uses allowed_distance
        )
      );

      return { success: true, data: stores };
    } catch (error) {
      console.error('StoreRepository.getStores error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to fetch stores',
      };
    }
  }

  /**
   * 새 상점 생성
   */
  async createStore(
    companyId: string,
    storeName: string,
    address: string | null,
    phone: string | null
  ) {
    try {
      await this.dataSource.createStore(companyId, storeName, address, phone);
      return { success: true };
    } catch (error) {
      console.error('StoreRepository.createStore error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to create store',
      };
    }
  }

  /**
   * 상점 삭제 (Soft Delete)
   */
  async deleteStore(storeId: string) {
    try {
      await this.dataSource.deleteStore(storeId);
      return { success: true };
    } catch (error) {
      console.error('StoreRepository.deleteStore error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to delete store',
      };
    }
  }
}
